### PR TITLE
Namespacing: Datadog::Tracing internals

### PR DESCRIPTION
**Pull request by delner**
This (massive) pull request moves all of the tracing components from `Datadog` in `lib/ddtrace` to `Datadog::Tracing` in `lib/datadog/tracing`.

There are a few exceptions, like `Transport` (used by both tracing and profiling, to be replaced later), `AutoInstrument`, `OpenTracer`, `OpenTelemetry`, and `Contrib`.

Almost all the changes are namespace updates, but there's also moving constants in `ddtrace/ext` to `datadog/tracing/metadata/ext` and `datadog/tracing/configuration/ext` where appropriate. Additionally, some files were split up or re-organized (like `DistributedTracing` and `Sampling`.)

### CODE REVIEW

File: ddtrace.gemspec [hunk 1]
```diff
@@ -6,8 +6,8 @@ require 'ddtrace/version'
 
 Gem::Specification.new do |spec|
   spec.name                  = 'ddtrace'
-  spec.version               = Datadog::VERSION::STRING
-  spec.required_ruby_version = [">= #{Datadog::VERSION::MINIMUM_RUBY_VERSION}", "< #{Datadog::VERSION::MAXIMUM_RUBY_VERSION}"]
+  spec.version               = DDTrace::VERSION::STRING
```
Comment (lloeki) [inline ddtrace.gemspec:9]: I see this making sense but I have an awkward feeling introducing a `DDTrace` constant with all the effort to move stuff towards `Datadog`.

File: lib/datadog/tracing.rb [hunk 2]
```diff
@@ -86,54 +87,54 @@ def configuration
...
 
         # Configure application normally
         Datadog.send(:internal_configure, &wrapped_block)
       end
 
-      # (see Datadog::Tracer#active_trace)
+      # (see Datadog::Tracing::Tracer#active_trace)
       # @public_api
-      def active_trace
+      def active_trace(key: nil)
```
Comment (marcotc) [inline lib/datadog/tracing.rb:99]: I don't think we should expose the `key` argument in the public API.

There's no valid use case for a `ddtrace` user that I can think of that can be achieve by providing a custom key here.

Key is a misnomer, as it is actually the storage vessel for the context (we only use it with a `Thread` object ourselves).

I think we should hide this parameter unless there's a clear need for it.

Same applies for other methods in this file with the `key` parameter added in this changeset.

File: spec/ddtrace/contrib/sidekiq/support/helper.rb [hunk 3]
```diff
@@ -51,6 +52,9 @@ def expect_in_sidekiq_server(duration: 2, &expectations)
       # spans created in the forked process, the exit status will be 0 even if our expectations fail.
       # Instead, look at `STDERR`, ignoring warnings.
       fork_expectations: proc do |status:, stdout:, stderr:|
+        stdout = Datadog::Core::Utils.utf8_encode(stdout) if stdout
+        stderr = Datadog::Core::Utils.utf8_encode(stderr) if stderr
```
Comment (marcotc) [inline spec/ddtrace/contrib/sidekiq/support/helper.rb]: Why were these added?

**Review by marcotc — APPROVED** [review]
Save for two small comments, all else looks good in this colossal PR: 
image

File: ddtrace.gemspec (hunk 1 above)
Comment (marcotc) [reply ddtrace.gemspec:9]: I think `DDTrace` looks strange, but it's very descriptive and accurate: it represents information tied to the gem itself, not to Datadog products.

Comment (marcotc) [reply ddtrace.gemspec:9]: (I'm sure open to suggestions, but I'm not against the current naming)

Comment (lloeki) [reply ddtrace.gemspec:9]: Yup that's exactly how I feel.

> it represents information tied to the gem itself, not to Datadog products.

What about `Datadog::Gem`? `Datadog::Gem::VERSION` has a nice ring to it. I'm thinking how in the future we plan to transition from `ddtrace` to `datadog`.

Comment (delner) [reply ddtrace.gemspec:9]: My strategic reason for doing this is that it allows us to introduce a `datadog` gem later without having a conflict with this one (Which will likely contain `Datadog::VERSION`.)

File: spec/ddtrace/contrib/sidekiq/support/helper.rb (hunk 3 above)
Comment (delner) [reply spec/ddtrace/contrib/sidekiq/support/helper.rb]: The tests were complaining that ASCII /UTF8 outputs were not compatible/comparable. I don't why this started to happen but this did fix the problem. (Maybe because it had some other error that triggered ASCII output?)

File: lib/datadog/tracing.rb (hunk 2 above)
Comment (delner) [reply lib/datadog/tracing.rb:99]: There's unfortunately a clear need for it: profiling. It needs to access spans/traces per thread so it can get correlation information. It can only do so from its own thread too.

**Review by marcotc — APPROVED** [review]

File: ddtrace.gemspec (hunk 1 above)
Comment (lloeki) [reply ddtrace.gemspec:9]: That's a quite reasonable rationale 👍
